### ISSUE TITLE
ssh-agent broken arm64 openssh 8.9

### ISSUE THREAD

**Issue by xboxones1** [keepassxreboot/keepassxc#7923]
## Overview
openssh 8.9. keepassxc 2.7.1 (older version/Snapshot) built from source. 
The keys are stored in keepassxc, after the database is unlocked, they are added to the ssh-agent. When executing the ssh-add -L command, the keys are displayed, when trying to clone a git repository via ssh, ssh-agent breaks and the repository is not cloned, when executing the ssh-add -L command, it writes "Error connecting to agent: Connection refused". The problem appeared after updating openssh to version 8.9. When rollback to OpenSSH 8.8p1 works fine

## Steps to Reproduce
1.  Add key to keepassxc
2.  After add key to ssh-agent
3.  Git clone repository via ssh

## Context
KeePassXC - Version 2.7.1
Revision: https://github.com/keepassxreboot/keepassxc/commit/5916a8f8dd93eb6a5de544caedd7a9d8e9a3b1ee

Qt 5.15.3
Debugging mode is disabled.

Operating system: Ubuntu 22.04 LTS
CPU architecture: arm64
Kernel: linux 5.15.0-1005

Enabled extensions:
    Auto-Type
    Browser Integration
    SSH Agent
    KeeShare
    YubiKey
    Secret Service integration

Cryptographic Libraries:
- Botan 2.19.1

**Comment by droidmonkey** [comment]
Open an issue with openssh, this isn't our bug. Once the keys are put into ssh-agent we are out of the picture. From your description that happens correctly.

**Comment by xboxones1** [comment]
But ssh-agent on arm works correctly without keepassxc. I don't understand what exactly the problem is, because everything works correctly on x86_64.

**Comment by droidmonkey** [comment]
Manually adding the same keys to ssh-agent and using git doesn't cause it to break with version 8.9?

What options are checked on the ssh agent page for those keys? What types of keys are they (RSA, EC)?

To be clear I am 100% certain this is an openssh problem, just teasing out some more info.

**Comment by xboxones1** [comment]
The key is tested by the same. Type key ed25519. When manually adding the key with the ssh-add id_ed25519 command, everything works fine, the repository is cloned. When using keepassxc ssh-agent just breaks and nothing works.

**Comment by droidmonkey** [comment]
We pass the keys using their API, not via the command line. They could have introduced a bug in processing keys from the API interface.

**Comment by droidmonkey** [comment]
New protocol @hifi ?

https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.agent

**Comment by hifi** [comment]
From the top of my head I have no idea why a new extension would break the existing protocol but apparently so.

**Comment by hifi** [comment]
I can't reproduce this with OpenSSH 9.0p1 on x86_64 Alpine against KeePassXC 2.7.1. The keys are loaded just fine to agent and the agent also works just fine when logging into somewhere with ssh.

```
...
debug1: process_add_identity: add ssh-ed25519 SHA256:... "hifi@keepassxc" (life: 0) (confirm: 0) (provider: none) (destination constraints: 0)
...
debug3: identity_permitted: entering: key ED25519 comment "hifi@keepassxc", 0 socket bindings, 0 constraints
debug2: process_request_identities: replying with 1 allowed of 1 available keys
```

Notice the new debug "(destination constraints: 0)" so the agent is definitely running a version that supports the new constraints.

I've also used KeePassXC quite a bit on an arm64 Arch Linux installation so arm64 itself shouldn't be the problem. Is your Raspbian build running in little endian or big endian mode as I haven't tested big endian at all so that might be broken anyway.

**Comment by xboxones1** [comment]
```
user@mate:~$  ssh -vvv root@192.168.1.1
OpenSSH_8.9p1 user-3, OpenSSL 3.0.2 15 Mar 2022
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: resolve_canonicalize: hostname 192.168.1.1 is address
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> '/home/user/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> '/home/user/.ssh/known_hosts2'
debug3: ssh_connect_direct: entering
debug1: Connecting to 192.168.1.1 [192.168.1.1] port 22.
debug3: set_sock_tos: set socket 3 IP_TOS 0x10
debug1: Connection established.
debug1: identity file /home/user/.ssh/id_rsa type -1
debug1: identity file /home/user/.ssh/id_rsa-cert type -1
debug1: identity file /home/user/.ssh/id_ecdsa type -1
debug1: identity file /home/user/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/user/.ssh/id_ecdsa_sk type -1
debug1: identity file /home/user/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /home/user/.ssh/id_ed25519 type -1
debug1: identity file /home/user/.ssh/id_ed25519-cert type -1
debug1: identity file /home/user/.ssh/id_ed25519_sk type -1
debug1: identity file /home/user/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /home/user/.ssh/id_xmss type -1
debug1: identity file /home/user/.ssh/id_xmss-cert type -1
debug1: identity file /home/user/.ssh/id_dsa type -1
debug1: identity file /home/user/.ssh/id_dsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.9p1 user-3
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0
debug1: compat_banner: match: OpenSSH_9.0 pat OpenSSH* compat 0x04000000
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to 192.168.1.1:22 as 'root'
debug1: load_hostkeys: fopen /home/user/.ssh/known_hosts: No such file or directory
debug1: load_hostkeys: fopen /home/user/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug3: order_hostkeyalgs: no algorithms matched; accept original
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug2: local client KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,sntrup761x25519-sha512@openssh.com,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group14-sha256,ext-info-c
debug2: host key algorithms: ssh-ed25519-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,sk-ssh-ed25519-cert-v01@openssh.com,sk-ecdsa-sha2-nistp256-cert-v01@openssh.com,rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-ed25519,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ssh-ed25519@openssh.com,sk-ecdsa-sha2-nistp256@openssh.com,rsa-sha2-512,rsa-sha2-256
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos: 
debug2: languages stoc: 
debug2: first_kex_follows 0 
debug2: reserved 0 
debug2: peer server KEXINIT proposal
debug2: KEX algorithms: curve25519-sha256@libssh.org,diffie-hellman-group14-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha256,sntrup761x25519-sha512@openssh.com
debug2: host key algorithms: ssh-ed25519
debug2: ciphers ctos: chacha20-poly1305@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com
debug2: MACs ctos: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128-etm@openssh.com
debug2: MACs stoc: hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,umac-128-etm@openssh.com
debug2: compression ctos: none,zlib@openssh.com
debug2: compression stoc: none,zlib@openssh.com
debug2: languages ctos: 
debug2: languages stoc: 
debug2: first_kex_follows 0 
debug2: reserved 0 
debug1: kex: algorithm: curve25519-sha256@libssh.org
debug1: kex: host key algorithm: ssh-ed25519
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug3: send packet: type 30
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug3: receive packet: type 31
debug1: SSH2_MSG_KEX_ECDH_REPLY received
debug1: Server host key: ssh-ed25519 SHA256:uFuKelKlkJoUKi/uio9NQqkn3nY4xgWwSoPItREi62
debug1: load_hostkeys: fopen /home/user/.ssh/known_hosts: No such file or directory
debug1: load_hostkeys: fopen /home/user/.ssh/known_hosts2: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts: No such file or directory
debug1: load_hostkeys: fopen /etc/ssh/ssh_known_hosts2: No such file or directory
debug3: hostkeys_find_by_key_hostfile: trying user hostfile "/home/user/.ssh/known_hosts"
debug1: hostkeys_find_by_key_hostfile: hostkeys file /home/user/.ssh/known_hosts does not exist
debug3: hostkeys_find_by_key_hostfile: trying user hostfile "/home/user/.ssh/known_hosts2"
debug1: hostkeys_find_by_key_hostfile: hostkeys file /home/user/.ssh/known_hosts2 does not exist
debug3: hostkeys_find_by_key_hostfile: trying system hostfile "/etc/ssh/ssh_known_hosts"
debug1: hostkeys_find_by_key_hostfile: hostkeys file /etc/ssh/ssh_known_hosts does not exist
debug3: hostkeys_find_by_key_hostfile: trying system hostfile "/etc/ssh/ssh_known_hosts2"
debug1: hostkeys_find_by_key_hostfile: hostkeys file /etc/ssh/ssh_known_hosts2 does not exist
The authenticity of host '192.168.1.1 (192.168.1.1)' can't be established.
ED25519 key fingerprint is SHA256:uFuKelKlkJoUKi/uio9NQqkn3nY4xgWwSoPItREi62.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '192.168.1.1' (ED25519) to the list of known hosts.
debug3: send packet: type 21
debug2: ssh_set_newkeys: mode 1
debug1: rekey out after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug3: receive packet: type 21
debug1: SSH2_MSG_NEWKEYS received
debug2: ssh_set_newkeys: mode 0
debug1: rekey in after 134217728 blocks
debug2: get_agent_identities: ssh_agent_bind_hostkey: communication with agent failed
debug1: get_agent_identities: ssh_fetch_identitylist: communication with agent failed
debug1: Will attempt key: /home/user/.ssh/id_rsa 
debug1: Will attempt key: /home/user/.ssh/id_ecdsa 
debug1: Will attempt key: /home/user/.ssh/id_ecdsa_sk 
debug1: Will attempt key: /home/user/.ssh/id_ed25519 
debug1: Will attempt key: /home/user/.ssh/id_ed25519_sk 
debug1: Will attempt key: /home/user/.ssh/id_xmss 
debug1: Will attempt key: /home/user/.ssh/id_dsa 
debug2: pubkey_prepare: done
debug3: send packet: type 5
debug3: receive packet: type 7
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com,webauthn-sk-ecdsa-sha2-nistp256@openssh.com>
debug1: kex_input_ext_info: publickey-hostbound@openssh.com=<0>
debug3: receive packet: type 6
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug3: send packet: type 50
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey
debug3: start over, passed a different list publickey
debug3: preferred gssapi-with-mic,publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Trying private key: /home/user/.ssh/id_rsa
debug3: no such identity: /home/user/.ssh/id_rsa: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_ecdsa
debug3: no such identity: /home/user/.ssh/id_ecdsa: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_ecdsa_sk
debug3: no such identity: /home/user/.ssh/id_ecdsa_sk: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_ed25519
debug3: no such identity: /home/user/.ssh/id_ed25519: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_ed25519_sk
debug3: no such identity: /home/user/.ssh/id_ed25519_sk: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_xmss
debug3: no such identity: /home/user/.ssh/id_xmss: No such file or directory
debug1: Trying private key: /home/user/.ssh/id_dsa
debug3: no such identity: /home/user/.ssh/id_dsa: No such file or directory
debug2: we did not send a packet, disable method
debug1: No more authentication methods to try.
root@192.168.1.1: Permission denied (publickey).

```

**Comment by xboxones1** [comment]
@hifi I have Ubuntu 22.04 LTS installed, both on PC and on Raspberry pi 4. Everything works fine on PC, the problem only occurred on raspberry pi. On raspberry I tried to build a new version of openssh 9.0 and there is the same behavior, ssh-agent breaks. Having searched about this problem, you can see if it occurs on different platforms and distributions, there are no specifics anywhere, but everywhere they write related to some kind of new protocol in openssh 8.9 and newer.

**Comment by hifi** [comment]
If `ssh-agent` _crashes_ on use then that is definitely a bug in OpenSSH for it to allow that to happen regardless what invalid data it's being fed.

**Comment by droidmonkey** [comment]
That was my original position on the matter

**Comment by xboxones1** [comment]
@hifi I gave an example with a keepass plugin https://github.com/dlech/KeeAgent/issues/341 and it was fixed there

**Comment by hifi** [comment]
The issue with KeeAgent seems to be about it running in agent mode which we don't implement at all. As you witnessed it works fine on your PC with the same version of OpenSSH so this is something else.

Can you reproduce the issue if you generate a new ed25519 key for testing to rule out your existing key?

**Comment by xboxones1** [comment]
It did not help, at the same time I tried the rsa key, nothing has changed.

**Comment by xboxones1** [comment]
@hifi After restarting ssh-agent, the $SSH_AUTH_SOCK variable will change. It starts working with the new variable if you specify it in the keepassxc settings.

**Comment by hifi** [comment]
That just seems to indicate the issue is with ssh-agent rather than KeePassXC. Maybe it's a constraint issue that it tries to prompt for a confirmation and that fails horribly? Do you have confirmation constraint set on KeePassXC but not when you add the key manually?

**Comment by xboxones1** [comment]
I don't understand what restrictions are in question, I have the same settings on both PC and raspberry. I didn't restrict anything. With bugtrackers, I don’t know where to write, I tried to contact the Ubuntu developers, but they didn’t find problems with ssh-agent there, when I wrote about keepassxc there it’s just silent. I don't know how to properly format and convey what the problem is. I think the openssh developers will also ignore it. I have already tried reinstalling ubuntu to a clean one. I think a simple solution would be to rebuild the openssh 8.8 package by overestimating the package version so that it is not updated.
